---
title: "Analysis of the patterns in inversions location"
output:
  html_document:
    df_print: paged
  pdf_document: default
bibliography: report/20220129_LocationPatterns/references.bib 
editor_options: 
  chunk_output_type: inline
---



<!-- # Objectives -->

<!-- Inversions can be classified in two groups depending on their generation process: two relatively close breaks that are repaired in opposite orientations by non-homologous mechanisms (which we call NH inversions), or non-allelic homologous recombination between inverted repeats (NAHR inversions). These two groups have differences regarding important characteristics such as size, location  and divergence between orientations. NH inversions are smaller, can appear anywhere in the genome and are unique, which causes a marked divergence between opposite orientations as time goes by. NAHR inversions appear under the specific conditions of having inverted repeats at a certain distance, but can be much bigger than NH inversions and can be generated more than once throughout the history of a population, which leads to less divergence between orientations [@Aguado2014; @Giner-Delgado2019]. -->

<!-- In addition, inversions generate aberrant chromosomes and unviable zygotes if a recombination event takes place within the region during meiosis. Thus, they are expected to have better chances of reaching polymorphism in low recombination regions. On the other hand, a higher incidence of NAHR events can be expected in high recombination regions [@Sasaki2010].  -->

<!-- In this notebook, I explore whether inversions are evenly distributed among chromosomes and which variables affect the generation and/or maintenance of inversions as polymorphisms in certain regions of the genome.  -->

```{r, eval = F}
if (!require("pacman")) install.packages("pacman")
#
# library(ggplot2)
# library(ggbio)
# library(GenomicRanges)
# library(reshape2)
# library(patchwork)
# data(ideoCyto, package = "biovizBase")
# library(performance)#
```


# Fine-scale patterns

```{r finescaleSetup}

# Setup
today <- "defTest"
gatfiledir<-paste0("analysis/",today,"_LocationPatterns/GATfiles/" )
gatresdir<-paste0("analysis/",today,"_LocationPatterns/GATresults/" )
# Frequency threshold
freq <- 0.2
Mb<-1.5
# Directory
dir.create(gatfiledir, recursive = T, showWarnings = F)
dir.create(gatresdir, recursive = T, showWarnings = F)

save(today, gatfiledir, gatresdir, freq, Mb, file="analysisparams.Rdata")
```




## Description of variables

```{r child = 'LocationPatterns_01_Description.Rmd'}

```

## Raw data visualization

```{r  child = 'LocationPatterns_00_Visualization.Rmd'}

```


## GAT run & Result

```{r child = 'LocationPatterns_02_Analysis.Rmd'}

```

# Adjacent test

```{r finescaleSetup}

# Setup
today <- "defTest"
gatfiledir<-paste0("analysis/",today,"_LocationPatterns/GATfiles/" )
gatresdir<-paste0("analysis/",today,"_LocationPatterns/GATresults/" )
# Frequency threshold
freq <- 0.2
Mb<-1.5
# Directory
dir.create(gatfiledir, recursive = T, showWarnings = F)
dir.create(gatresdir, recursive = T, showWarnings = F)

save(today, gatfiledir, gatresdir, freq, Mb, file="analysisparams.Rdata")
```

```{bash}

TODAY="defTest"
 
for file in $(ls analysis/${TODAY}_LocationPatterns/GATfiles/Segments/*); do
  DEST="$(echo $file | sed 's/.bed//g')_sides.bed"
  awk -v OFS="\t" 'l=$3-$2 {print $1,  $2-l, $2, $4 "_left"}' $file  > $DEST
  awk -v OFS="\t" 'l=$3-$2 {print $1,  $3, $3+l, $4 "_right"}' $file  >> $DEST
done

```

```{r child = 'LocationPatterns_02_Analysis_adjacent.Rmd'}
# This needs some changes to work
```

```{bash}

TODAY="defTest"
 
for file in $(ls analysis/${TODAY}_LocationPatterns/GATfiles/Segments/*invs.bed); do
  DEST="$(echo $file | sed 's/.bed//g')_doublesides.bed"
  awk -v OFS="\t" 'l=2*($3-$2) {print $1,  $2-l, $2, $4 "_left"}' $file  > $DEST
  awk -v OFS="\t" 'l=2*($3-$2) {print $1,  $3, $3+l, $4 "_right"}' $file  >> $DEST
done

```

```{r child = 'LocationPatterns_02_Analysis_adjacent.Rmd'}
# This needs some changes to work

```




# How is each map affected

```{r child = 'LocationPatterns_03_SeichBasatena.Rmd'}

```

