---
title: "Analysis of the patterns in inversions location"
output:
  html_document:
    df_print: paged
  pdf_document: default
bibliography: report/20220129_LocationPatterns/references.bib 
editor_options: 
  chunk_output_type: inline
---



# Objectives

Inversions can be classified in two groups depending on their generation process: two relatively close breaks that are repaired in opposite orientations by non-homologous mechanisms (which we call NH inversions), or non-allelic homologous recombination between inverted repeats (NAHR inversions). These two groups have differences regarding important characteristics such as size, location  and divergence between orientations. NH inversions are smaller, can appear anywhere in the genome and are unique, which causes a marked divergence between opposite orientations as time goes by. NAHR inversions appear under the specific conditions of having inverted repeats at a certain distance, but can be much bigger than NH inversions and can be generated more than once throughout the history of a population, which leads to less divergence between orientations [@Aguado2014; @Giner-Delgado2019].

In addition, inversions generate aberrant chromosomes and unviable zygotes if a recombination event takes place within the region during meiosis. Thus, they are expected to have better chances of reaching polymorphism in low recombination regions. On the other hand, a higher incidence of NAHR events can be expected in high recombination regions [@Sasaki2010]. 

In this notebook, I explore whether inversions are evenly distributed among chromosomes and which variables affect the generation and/or maintenance of inversions as polymorphisms in certain regions of the genome. 

```{r, eval = F}
if (!require("pacman")) install.packages("pacman")
#
# library(ggplot2)
# library(ggbio)
# library(GenomicRanges)
# library(reshape2)
# library(patchwork)
# data(ideoCyto, package = "biovizBase")
# library(performance)#
```


# Large-scale patterns

## Workspace and windows definition 

The term 'workspace' refers to the genomic regions suitable for analysis. To define it, we calculated the first and last coordinates from each chromosome arm to have estimations for the recombination map being analyzed. We excluded the windows overlapping with the centromere, which are usually very large due to the lower data availability when estimating recombination maps, and thus we considered them to be less reliable. 

Once the workspace is defined, each arm is divided into an equal number of windows, generating window sizes proportional to arm size that allow us to have physical size as a variable. In addition, for each window we calculate inversion center point overlap, repeat center point overlap, and mean recombination rate (both simple and map window size weighted means).

There are two types of windows: fixed size windows and crossover zones. Fixed size windows are calculated dividing each chromosome arm (centromere excluded) into n windows. Crossover zones are the spaces between two local minima in the crossover density. Since we only have recombination rate values, we resample the coordinates into what can be considered a simulation of crossovers, and then calculate crossover zones similarly to Bell et al. 2020. 

```{bash}
ENVPY="/home/rgomez/anaconda3/bin/python"
# ANALYSIS WIDOWS: 5 fixed windows, and crossover zones with 0 persistence and 40k simulations (sample size similar to Bell et al.2020)

# Spence CEU map
$ENVPY code/python/divideChromosomes.py -m data/SpenceSong_hg19_recMaps_processed/CEU_recombination_map_hg19_allChr.bed -b data/cytoBand.txt -f 5 -s 0.1 -r 800000 -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d CEUSpence
# Bherer map (general)
$ENVPY code/python/divideChromosomes.py -m data/Bherer_Refined_genetic_map_b37_processed/sexavg_allChr.bed -b data/cytoBand.txt -f 5 -s 0.1 -r 800000 -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d avgBherer 
# Bherer map (female)
$ENVPY code/python/divideChromosomes.py -m data/Bherer_Refined_genetic_map_b37_processed/female_allChr.bed -b data/cytoBand.txt -f 5 -s 0.1 -r 800000 -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d femBherer

# ------------ TEST

$ENVPY code/python/divideChromosomes.py -m data/SpenceSong_hg19_recMaps_processed/CEU_recombination_map_hg19_1481711.bed -b data/cytoBand.txt -f 5 -s 0.1 -r 100000 -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d CEUSpence_part
```

```{bash valleypeaks}
TODAY=$(date +'%Y%m%d')
USE=20220513

mkdir -p "analysis/$(date +'%Y%m%d')_LocationPatterns/divideChromosomes"


cp -r "analysis/${USE}_LocationPatterns/divideChromosomes/CEUSpence_COzones_0.1_800000" "analysis/$(date +'%Y%m%d')_LocationPatterns/divideChromosomes/CEUSpence_VPzones_0.1_800000"

cp -r "analysis/${USE}_LocationPatterns/divideChromosomes/avgBherer_COzones_0.1_800000" "analysis/$(date +'%Y%m%d')_LocationPatterns/divideChromosomes/avgBherer_VPzones_0.1_800000"

cp -r "analysis/${USE}_LocationPatterns/divideChromosomes/femBherer_COzones_0.1_800000" "analysis/$(date +'%Y%m%d')_LocationPatterns/divideChromosomes/femBherer_VPzones_0.1_800000"

```

```{r valleypeaks}
date <- 20220516


for (name in c("avgBherer", "femBherer", "CEUSpence")) {
  
  windows<-data.frame()
  extremes<-read.table(paste0("analysis/",date,"_LocationPatterns/divideChromosomes/",name,"_VPzones_0.1_800000/extremes.txt"), sep = "\t", header = T)
  
  extremes<-extremes[order(extremes$Chromosome, extremes$pos),]
  
  
  for (chr in unique(extremes$Chromosome)) {
    part<-extremes[extremes$Chromosome == chr,]
    part$pos2<- with(part, c(pos[2:length(pos)],pos[length(pos)])  )
    part$center<-with(part, pos+(pos2-pos)/2)
    coords<-with(part, c(pos[1], center[2:(length(center)-2)], pos[length(pos)]))
    starts<-coords[1:length(coords)-1]
    ends<-coords[2:length(coords)]
    windows<- rbind(windows, data.frame(Start = starts, End = ends, Chromosome = chr))
  }
  
  write.table(windows, paste0("analysis/",date,"_LocationPatterns/divideChromosomes/",name,"_VPzones_0.1_800000/windows.txt"), sep = "\t", col.names = T, row.names = F,  quote = F )

}
```

## Window information
For each window, we measure multiple variables related to repeats, crossover rate, inversions...

```{bash}

ENVPY="/home/rgomez/anaconda3/bin/python"
TODAY=$(date +'%Y%m%d')
USE=$TODAY

for DIR in $(ls -d analysis/${USE}_LocationPatterns/divideChromosomes/*/); do

RECMAP=$(grep "recombination map" ${DIR}/log.txt | cut -d ":" -f2)
NAME=$(basename ${DIR})

$ENVPY code/python/fillWindowMeasurements.py -m ${RECMAP} -w ${DIR}/windows.txt  -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d $NAME

done

```

## Variable exploration and model exploration
```{bash statAnatalysis}

TODAY=$(date +'%Y%m%d')
USE=20220513

mkdir -p "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis"
echo "Starting proces..." > "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis/logfile.txt"

for DIR in $(ls -d analysis/${USE}_LocationPatterns/divideChromosomes/*/); do

echo $DIR >> "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis/logfile.txt"
NAME=$(basename ${DIR})

Rscript code/rscript/statAnalysis.R ${DIR} analysis/${USE}_LocationPatterns/fillWindowMeasurements/${NAME}  "analysis/${TODAY}_LocationPatterns" $NAME \
>> "analysis/${TODAY}_LocationPatterns/statAnalysis/logfile.txt" 2>&1
done

```

```{bash statAnatalysis_density}

TODAY=$(date +'%Y%m%d')
USE=20220513

mkdir -p "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis_density"
echo "Starting proces..." > "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis_density/logfile.txt"

for DIR in $(ls -d analysis/${USE}_LocationPatterns/divideChromosomes/*/); do

echo $DIR >> "analysis/$(date +'%Y%m%d')_LocationPatterns/statAnalysis_density/logfile.txt"
NAME=$(basename ${DIR})

Rscript code/rscript/statAnalysis_density.R ${DIR} analysis/${USE}_LocationPatterns/fillWindowMeasurements/${NAME}  "analysis/${TODAY}_LocationPatterns" $NAME \
>> "analysis/${TODAY}_LocationPatterns/statAnalysis_density/logfile.txt" 2>&1
done

```






# -------------------------- OLD ----------------------------#

```
python code/python/divideChromosomes.py -m data/Bherer_Refined_genetic_map_b37_processed/sexavg_allChr.bed -b data/cytoBand.txt -f 5 -s 0 -r 5000 -o "analysis/$(date +'%Y%m%d')_LocationPatterns"
```



```
python code/python/ArmSections.py -i data/InversionsAnnotation_131Inv_20211117.csv -r data/genomicSuperDups.txt -p data/Bherer_Refined_genetic_map_b37_procesed/BhererAllChroms.txt -l data/SpenceSong_hg19_recMaps_processed/CEU_SRR.bed -b data/cytoBand.txt -f 5 -o "analysis/$(date +'%Y%m%d')_LocationPatterns"
```


```{r  Load Data}
dateAnalysis <- "20220217"
dateArmSections <-"20220217"

theoryLimits <-read.table(paste0("analysis/",dateAnalysis,"_LocationPatterns/",dateArmSections,"_ArmSections/theoreticalChromosomeLimits.csv"), 
                          sep = "\t", header = T)
dataLimits <- read.table(paste0("analysis/",dateAnalysis,"_LocationPatterns/",dateArmSections,"_ArmSections/actualChromosomeLimits.csv"), 
                         sep = "\t", header = T)
windowData <- read.table(paste0("analysis/",dateAnalysis,"_LocationPatterns/",dateArmSections,"_ArmSections/windowData.csv"), 
                         sep = "\t", header= T)
```

```{r Show effective workspace, cache = TRUE}
mergedLimits <- merge(theoryLimits, dataLimits, by = "chromID", suffixes = c("_theory", "_data") , all = TRUE)
mergedLimits <- mergedLimits[complete.cases(mergedLimits),]

# ARM REGIONS 
arms.GR<-makeGRangesFromDataFrame(mergedLimits, ignore.strand = T, start.field = "Start_data", end.field = "End_data" , seqnames.field = "chrom")

# ORIGIN OF LIMITS
viewWidth = 100000
mergedLimits$Start_data_end <- mergedLimits$Start_data + viewWidth
mergedLimits$End_data_start <- mergedLimits$End_data - viewWidth

limits.GR <-makeGRangesFromDataFrame(
  rbind(
  setNames(data.frame(mergedLimits[ ,c("chrom", "Start_data","Start_data_end",  "sourceStart")]), c("chrom","Start","End", "Source")) ,
  setNames(data.frame(mergedLimits[,c("chrom", "End_data_start", "End_data", "sourceEnd")]), c("chrom","Start", "End","Source"))
  ),
  keep.extra.columns = TRUE)

autoplot(ideoCyto$hg19, layout = "karyogram", cytobands = TRUE)+
  layout_karyogram(data = arms.GR, geom = "rect", ylim = c(0, 10), fill = "#91b5b5", alpha = 0.5)+
  layout_karyogram(data = limits.GR, geom = "rect", ylim = c(0, 10), aes(color = Source), alpha = 0, size = 1)+
  guides(fill = FALSE)+
  scale_color_manual("Source", values =c("Bherer" = "purple", "Spence" = "green"))+
  ggtitle("Effective workspace")

```


## Exploration of variables

```{r VariableVisualization, fig.width=11, fig.heigth=6, dpi = 300, cache = TRUE}

# windowData
windowData$Chromosome <- factor(windowData$Chromosome, levels = paste(rep("chr", 22), as.character(c(1:22)), sep = ""))

windowDataMelted <- melt(windowData, id.vars = c("chromID", "winID", "Chromosome", "Start", "End")) 

windowDataMelted$dataGroup<- ifelse(windowDataMelted$variable %in% c("invCenters", "NHCenters", "NAHRCenters"), "Inversions", 
                             ifelse(windowDataMelted$variable %in% c("allRepCounts", "intraRepCounts"), "Repeats",
                             ifelse(windowDataMelted$variable %in% c("BhererRegsNum", "SpenceRegsNum"), "recWindows",
                             ifelse(windowDataMelted$variable %in% c("BhererRegsMean", "SpenceRegsMean", "BhererRegsWeightMean",
                                                                     "SpenceRegsWeightMean"), "recMeans", "Length"))))

# Add log repeats
windowDataMelted_replog <- windowDataMelted[windowDataMelted$dataGroup == "Repeats",]
windowDataMelted_replog$value <- log10(windowDataMelted_replog$value)
windowDataMelted_replog$dataGroup <- "Repeats_log10"
windowDataMelted<- rbind(windowDataMelted, windowDataMelted_replog)


plot_list<-list()
for (group in unique(windowDataMelted$dataGroup)) {
  
  plotTable <- windowDataMelted[(windowDataMelted$value != -Inf) & (windowDataMelted$dataGroup == group),]
  plot_list[[group]] <- ggplot(plotTable, aes(x = variable, y = value))+
    # Half violin
      ggdist::stat_halfeye(adjust = .5, width = .6, .width = 0, justification = -.2, point_colour = NA) +
    # Boxplot 
      geom_boxplot(width = .1, outlier.shape = NA) +
    # Points
      gghalves::geom_half_point_panel(side = "l", range_scale = .6,  alpha = .5, aes(color = Chromosome))+
      scale_color_manual(values = c(rep("#3c7ae7",11),rep("#89b23e",11) ))+
    # Adjust coordinates
      coord_flip()+
      # coord_flip( xlim = c(1.3, NA))+
    # Adjust labels
      theme(axis.title.y = element_blank(), legend.position = "none")+
    # Title
      ggtitle(group)
}
```



```{r VariableVisualizationPlot, fig.height=11, fig.width=20, cache = TRUE}

wrap_plots(plot_list)
```

```{r VariableVisualizationPlot, fig.height=11, fig.width=20, cache = TRUE}
c<-table(windowData$NHCenters, windowData$NAHRCenters )
names(dimnames(c))<-c("NHCenters", "NAHRCenters")
c

windowData$chromSection <- ifelse(grepl("p0|q4",windowData$winID), "Telomere", "Arm")

c<-table(windowData$invCenters , windowData$chromSection)
names(dimnames(c))<-c("invCenters", "chromSection")
c
```

```{r examples}
#-----------------------------#
# # Plot
# # Function to calculate number of observations
# n_fun <- function(x){
#   return(data.frame(y = median(x)*1.5, 
#                     label = paste0("n = ",length(x))))
# }
# 
# library(ggdist)
# library(viridis)
# ggplot(windowDataMelted[(windowDataMelted$value != -Inf) & (windowDataMelted$dataGroup %in% c("Inversions")),], aes(x = variable, y = value ))+
#   # facet_wrap(dataGroup+variable~., scales="free")+
#   # RAINCLOUD PLOT  # https://www.cedricscherer.com/2021/06/06/visualizing-distributions-with-raincloud-plots-and-how-to-create-them-with-ggplot2/#back2
#     # Half violin
#       ggdist::stat_halfeye(adjust = .5, width = .6, .width = 0, justification = -.15, point_colour = NA) + 
#     # Boxplot 
#       geom_boxplot(width = .10, outlier.shape = NA) +
#     # Points (1 line = 1 option)
#       # geom_point(size = 1.3,alpha =.5,position = position_jitter(seed =  1, width = .04) , aes(color = Chromosome) ) + # this works for chormosomes
#       # ggdist::stat_dots(side = "left", justification = 1.1, binwidth = 3e+05  )+ # rain
#       # gghalves::geom_half_point(side = "l", range_scale = .5, alpha = .3)+
#       # geom_point(shape = 124 ,size = 10,alpha = .2, aes(color = Chromosome)) + # shape 124 in horizontal and 95 in vertical, this works for chormosomes
#   # Adjust coordinates
#     coord_flip( xlim = c(1.3, NA))+
#     # coord_flip()+
#   # Adjust colors
#     # scale_color_viridis(discrete = TRUE)
#     scale_color_manual(values = c(rep("#3c7ae7",11),rep("#89b23e",11) ) )+
#   # Title
#     ggtitle("Raincloud plot")
#     
```

```{r distrbutionFitting}

# Do I need this? 

# library(fitdistrplus)
# library(logspline)
# 
# 
# fit_list<-list()
# for (group in unique(windowDataMelted$variable)) {
#   
# descdist(windowData$BhererRegsWeightMean, discrete =  ifelse(grepl("Mean" , group ), FALSE, TRUE), boot = 1000)
# }
# 


```

```{r correlations, fig.height=10, fig.width=10, cache = TRUE}

library("corrplot")
corrTable<-windowData[,c("NHCenters", "NAHRCenters", "Length.bp.", "allRepCounts", "BhererRegsWeightMean") ]
corrTable$log10allRepCounts<-log10(corrTable$allRepCounts)

rownames(corrTable) <- paste0(windowData$winID)

# QUICK EASY ALTERNATIVE
# toPlotCor = cor(as.matrix(corrTable), method = "pearson")
# testRes = cor.mtest(corrTable, conf.level = 0.95)
# 
# corrplot(toPlotCor, p.mat = testRes$p, sig.level = c(0.001, 0.01, 0.05), pch.cex = 0.9, insig = 'label_sig', type = "lower", method = "square", col = COL2('PRGn'), diag = FALSE, tl.col = "black")

library("PerformanceAnalytics")
chart.Correlation(corrTable, histogram=TRUE, pch=19, method = "pearson")
mtext("Pearson correlation", line = 1)
chart.Correlation(corrTable, histogram=TRUE, pch=19, method = "spearman")
mtext("Spearman correlation", line = 1)
```

## Multiple linear regression

```{r tutorialModel}
# We will use the variables inversions (all), repeats (all), weighted mean, window length

##### Relationship between independent variables
# ggplot(lmData)+aes(x = Length.bp., y = NAHRCenters, size = BhererRegsWeightMean, color = allRepCounts)+geom_point() + scale_color_gradient()

# Make model with RepCounts
model<-lm(log10(allRepCounts) ~   BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)

summary(model)
```

```{r tutorialModel}
# Make model with logRepCounts
model<-lm(log10(allRepCounts) ~   BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
##### Add Inversions

model<-lm(invCenters ~  allRepCounts + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
model<-lm(invCenters ~   log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
##### Separate inversions - NH
model<-lm(NHCenters ~  allRepCounts + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
model<-lm(NHCenters ~   log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
model<-lm(NAHRCenters ~  allRepCounts + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```

```{r tutorialModel}
model<-lm(NAHRCenters ~   log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)
summary(model)

check_model(model)
```
```{r stepAICinv}

fit<-lm(invCenters ~   allRepCounts+ log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)

step <- stepAIC(fit, direction="both") # from MASS library
step$anova # display results 

model <- lm(invCenters ~ allRepCounts + Length.bp., data = windowData)
summary(model)
check_model(model)
```

```{r stepAICNH}

fit<-lm(NHCenters ~   allRepCounts+ log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)

step <- stepAIC(fit, direction="both") # from MASS library
step$anova # display results 

model <- lm(NHCenters ~ Length.bp. + as.factor(chromSection), data = windowData)
summary(model)
check_model(model)
```

```{r stepAICNAHR}

fit<-lm(NAHRCenters ~   allRepCounts+ log(allRepCounts) + BhererRegsWeightMean +  Length.bp.  + as.factor(chromSection), data = windowData)

step <- stepAIC(fit, direction="both") # from MASS library
step$anova # display results 

model <- lm(NAHRCenters ~ allRepCounts + BhererRegsWeightMean + as.factor(chromSection), data = windowData)
summary(model)
check_model(model)
```

```{r tutorialModel}
##### ORDINAL LINEAR REGRESSION - https://www.r-bloggers.com/2019/06/how-to-perform-ordinal-logistic-regression-in-r/
# Ordinal logistic models consider the probability of an event and all the events that are below the focal event in the ordered hierarchy.
# Independent variables can be measured on nominal, ordinal or continuous measurement scale, but must not be collinear.
mData<-windowData[,c( "invCenters",  "allRepCounts", "Length.bp.", "BhererRegsWeightMean", "chromSection")]
rownames(mData)<-windowData$winID
mData$invCenters <- factor(mData$invCenters, ordered = TRUE)
mData$chromSection<-factor(mData$chromSection,  ordered = TRUE, labels = c("0", "1"))
mData[,c("Length.bp.", "BhererRegsWeightMean", "allRepCounts")]<-sapply(mData[,c("Length.bp.", "BhererRegsWeightMean", "allRepCounts")], as.numeric)

#Dividing data into training and test set
#Random sampling 
samplesize = 0.60*nrow(mData)
set.seed(100)
index = sample(seq_len(nrow(mData)), size = samplesize)
#Creating training and test set 
datatrain = mData[index,]
datatest = mData[-index,]

#Build ordinal logistic regression model
library("MASS")
model= polr(invCenters ~ scale(allRepCounts) + scale(Length.bp.) + scale(BhererRegsWeightMean)+chromSection , data = datatrain, Hess = TRUE)

summary(model)

```

## Location along chromosome arm

```{r  Location along chromosome arm}

# Refresh data
dateAnalysis <- "20220217"
dateArmSections <-"20220217"

windowData <- read.table(paste0("analysis/",dateAnalysis,"_LocationPatterns/",dateArmSections,"_ArmSections/windowData.csv"), sep = "\t", header= T)

library("stringr")
windowData$chrpos<-str_sub(windowData$winID, -2 , -1)

optBtable<-aggregate(invCenters ~ chrpos, data=windowData, sum)
optBtable$opts<-table(windowData$chrpos)
optB<-ggplot(optBtable,aes(x=chrpos, y = invCenters/opts))+geom_bar( stat="identity")
  
# -----

# Take chromosome positions
chrSizes<-read.table("data/chrSizes.txt", head = T)
# Take inv positions
inv<-read.table("data/InversionsAnnotation_131Inv_20211117.csv", sep = "\t", header = TRUE, skip = 1, stringsAsFactors = F)
inv<-inv[,c("INV", "Origin", "Chr", "BP1_1.1", "BP2_2.1" )]
inv$Origin<- factor(inv$Origin, labels = c("NAHR", "NAHR", "NH", "NH"))
# Calculate positional percentage
inv$center<-inv$BP1_1.1 + (inv$BP2_2.1 - inv$BP1_1.1 +1)-1
inv<-merge(inv, chrSizes, by.x = "Chr", by.y ="Chromosome")
inv$percentagePos<-inv$center / inv$Totallength.bp.

optA<-ggplot(inv)+geom_histogram(aes(x = percentagePos), bins = 10)+facet_wrap(Origin  ~ .)

# --------
# Take arm positions
chrcen<-read.table("data/cytoBand.txt")
colnames(chrcen)<-c("Chr", "Start", "End", "band", "gstain")
chrcen$arm<-str_sub(chrcen$band, 1,1)
chrcen$arm<-paste0(chrcen$Chr, chrcen$arm)

chrcen<-merge(aggregate( Start ~ arm, chrcen, min ), aggregate( End ~ arm, chrcen, max ))

# Determine centromere point
chrcen$chr<-str_replace_all(chrcen$arm, c("p" = "", "q"= ""))

centromerepoint<-aggregate(End ~ chr, chrcen,min)

inv<-merge(centromerepoint, inv, by.y = "Chr", by.x = "chr")

inv$arm<-ifelse(inv$center <= inv$End,"p", "q")

inv$percentageArm<-ifelse(inv$arm == "p"  , inv$center / inv$End , 1-((inv$center-inv$End+1) / (inv$Totallength.bp. - inv$End +1)) )

summary(inv$percentageArm)
# 0 = telomere, 1 = centromere

optC<-ggplot(inv)+geom_histogram(aes(x = percentageArm), bins = 10)+facet_wrap(Origin ~ .)

# ---- plots
optA
optB
optC
```





## Large-scale patterns: functional sections

> Repetir el análsis de brazos cromosómicos (el del modelo) dividiendo cada brazo en 3 zonas (telomérica, central y centromérica) según un threshold de tasa de recombinación. Además, hacerlo con el mapa de Spence (histórico) en lugar del de Bherer (pedigrís), ya que utilizaremos el de Spence como el “estándar”. 

## Workspace and windows definition

1 - Check if the patterns from Avery can be observed here
  
```
python code/python/CrossoverZones.py -i data/InversionsAnnotation_131Inv_20211117.csv -r data/genomicSuperDups.txt -p data/Bherer_Refined_genetic_map_b37_procesed/BhererAllChroms.txt -l data/SpenceSong_hg19_recMaps_processed/CEU_recombination_map_hg19_allChr.bed -b data/cytoBand.txt -s 0.1 -o "analysis/$(date +'%Y%m%d')_LocationPatterns"
```

```{r}
library(ggplot2)
library(reshape2)

date1 <- "20220522"
date2 <- "20220522"
windowData <- read.table(paste0("analysis/",date1,"_LocationPatterns/",date2,"_CrossoverZones/COzone_windowData.csv"), header = T)

densities <-  read.table(paste0("analysis/",date1,"_LocationPatterns/",date2,"_CrossoverZones/densities.csv"), header = T)
extremes <-read.table(paste0("analysis/",date1,"_LocationPatterns/",date2,"_CrossoverZones/extremes.csv"), header = T)
histogram <- read.table(paste0("analysis/",date1,"_LocationPatterns/",date2,"_CrossoverZones/histogram.csv"), header = T)
histogram<-melt( histogram, id.vars = c("Start", "End","chromID", "winID", "Chromosome", "Length.bp."))
colnames(histogram)[colnames(histogram) %in% c("variable", "value")] <- c("Stat", "StatValue")

windowData$StatValue<-ifelse(windowData$Stat == "BhererRegsWeightMean", windowData$BhererRegsWeightMean, windowData$SpenceRegsWeightMean)
limits<-read.table(paste0("analysis/",date1,"_LocationPatterns/",date2,"_CrossoverZones/actualChromosomeLimits.csv"), header = T)
centromeres<-limits[grep("cen" , limits$chromID),]
centromeres$Chromosome<-sub("cen", "", centromeres$chromID)

ggplot()+
  geom_rect(data = histogram,aes( xmin = Start, xmax = End,  ymin = 0, ymax = StatValue ))+
  geom_line(data = densities,aes(x = pos, y = valScaled), size = 0.3, alpha = 0.8)+
  geom_point(data = extremes, aes(x = pos, y =valScaled, color = Type), size = 0.3)+
  geom_rect(data=windowData, aes(xmin = Start, xmax = End, fill = Color, ymin = 0, ymax = StatValue ), alpha = 0.4 , show.legend = F)+
  facet_wrap(Stat ~ Chromosome , scales = "free")+
  ggtitle("Crossover zones in recombination maps")


```
```{r}
ggplot()+
  geom_rect(data = histogram[histogram$Chromosome == "chr1",],aes( xmin = Start, xmax = End,  ymin = 0, ymax = StatValue ))+
  geom_line(data = densities[densities$Chromosome == "chr1",],aes(x = pos, y = valScaled), size = 0.5, alpha = 0.8)+
  geom_point(data = extremes[extremes$Chromosome == "chr1",], aes(x = pos, y =valScaled, color = Type), size = 0.5)+
  geom_rect(data=windowData[windowData$Chromosome == "chr1",], aes(xmin = Start, xmax = End, fill = Color, ymin = 0, ymax = StatValue ), alpha = 0.4 , show.legend = F)+
  facet_wrap(Stat ~ Chromosome , scales = "free")+
  ggtitle("Crossover zones in recombination maps")

```
```{r}
ggplot()+
  geom_rect(data = histogram[histogram$Chromosome == "chr4",],aes( xmin = Start, xmax = End,  ymin = 0, ymax = StatValue ))+
  geom_line(data = densities[densities$Chromosome == "chr4",],aes(x = pos, y = valScaled), size = 0.5, alpha = 0.8)+
  geom_point(data = extremes[extremes$Chromosome == "chr4",], aes(x = pos, y =valScaled, color = Type), size = 0.5)+
  geom_rect(data=windowData[windowData$Chromosome == "chr4",], aes(xmin = Start, xmax = End, fill = Color, ymin = 0, ymax = StatValue ), alpha = 0.4 , show.legend = F)+
  facet_wrap(Stat ~ Chromosome , scales = "free")+
  ggtitle("Crossover zones in recombination maps")


sc4<-read.table("data/SpenceSong_hg19_recMaps/YRI/YRI_recombination_map_hapmap_format_hg19_chr_4.txt", header = T)
scall<-read.table("data/SpenceSong_hg19_recMaps_processed/YRI_recombination_map_hg19_allChr.bed", header = F)
scall<- scall[scall$V1 == "chr4",]

bcall <- read.table("data/Bherer_Refined_genetic_map_b37_procesed/BhererAllChroms.txt", header = F) 
bcall<- bcall[bcall$V1 == "chr4",]

scall$V4 / sc4$Rate.cM.Mb.

plotable1<-sc4[sc4$Position.bp. > 37859896 & sc4$Position.bp. < 61675279,]
plotable2<-histogram[histogram$Stat == "SpenceRegsWeightMean" & histogram$Chromosome == "chr4" & histogram$Start > 37859896 & histogram$End < 61675279,]
plotable2$StatValue <- plotable2$StatValue / 1e-08
ggplot()+
  geom_rect(data =plotable2  , aes(xmin = Start, xmax = End, ymin = 0, ymax = StatValue))+
  geom_point(data = plotable1 ,aes(x = Position.bp., y = Rate.cM.Mb.), alpha = 0.1)

```

```{r}
library(ggplot2)
windowData<-read.table("analysis/20220404_LocationPatterns/20220404_ArmSections/windowData.csv", sep  ="\t", header = T)

chromnames<-c(paste0("chr", c(1:22)))

levels(windowData$Chromosome) <- chromnames

ggplot(windowData)+
  geom_rect(aes(xmin = Start, xmax = End, ymin= 0, ymax = SpenceRegsWeightMean ))+
  facet_wrap(Chromosome ~ ., scales = "free")

ggplot(windowData, aes(x = Start + (End-Start+1)/2, y = cumsum(SpenceRegsWeightMean)))+
  geom_line()+
  facet_wrap(Chromosome ~ ., scales = "free")

windowData$center<-with(windowData , Start + (End-Start+1)/2)

windowData[,c("cumulative", "previous", "previous_coord")]<-NA

for(c in chromnames){
  t <- windowData[windowData$Chromosome == c,]
  t$cumulative<-cumsum(t$SpenceRegsWeightMean)
  t$previous <- c(0,t$cumulative[1:length(t$cumulative)-1])
  t$previous_coord<- c(0,t$center[1:length(t$center)-1])
    
  windowData[windowData$Chromosome == c,]<-t
}

windowData$rise<-windowData$cumulative-windowData$previous
windowData$run<-windowData$center-windowData$previous_coord
windowData$slope<-windowData$rise / windowData$run

maxSlope<-aggregate(slope ~ Chromosome ,windowData, max)
colnames(maxSlope)<-c("Chromosome", "maxslope")
windowData<-merge(windowData, maxSlope, all = T)
windowData$normSlope<-windowData$slope / windowData$maxslope

library(viridisLite)
library(viridis)
ggplot(windowData, aes(x = center, y = SpenceRegsWeightMean, color = slope))+
  geom_line(size = 1)+
  scale_color_viridis()+
  facet_wrap(Chromosome ~ ., scales = "free")


ggplot(windowData, aes(x = center, y = cumulative, color = normSlope))+
  geom_bar(stat="identity")+
  scale_color_viridis()+
  facet_wrap(Chromosome ~ ., scales = "free")

maxcumulative<-aggregate(cumulative ~ Chromosome ,windowData, max)
colnames(maxcumulative)<-c("Chromosome", "maxcumsum")
windowData<-merge(windowData, maxcumulative, all = T)
windowData$normcumsum<-windowData$cumulative / windowData$maxcumsum

ggplot(windowData, aes(x = center, y = normcumsum, color = normSlope))+
  geom_bar(stat="identity")+
  scale_color_viridis()+
  facet_wrap(Chromosome ~ ., scales = "free")

ggplot(windowData)+geom_point(aes(x = normcumsum, y = normSlope))

#----- byarm

windowData <- read.table("analysis/20220404_LocationPatterns/20220404_ArmSections/windowData.csv", sep  ="\t", header = T)

chromnames<-c(paste0("chr", c(1:22)))
levels(windowData$Chromosome) <- chromnames

windowData$center<-with(windowData , Start + (End-Start+1)/2)
windowData[ grep("q", windowData$chromID) , "center"]<- windowData[ grep("q", windowData$chromID) , "center"]*-1

windowData[,c("cumulative", "previous", "previous_coord")]<-NA
windowData<-windowData[order(windowData$Chromosome, windowData$center),]

for(c in unique(windowData$chromID)){
  t <- windowData[windowData$chromID == c,]
  
  t$cumulative<-cumsum(t$SpenceRegsWeightMean)
  t$previous <- c(0,t$cumulative[1:length(t$cumulative)-1])
  t$previous_coord<- c(0,t$center[1:length(t$center)-1])

  windowData[windowData$chromID == c,]<-t
}

windowData$rise<-windowData$cumulative-windowData$previous
windowData$run<-windowData$center-windowData$previous_coord
windowData$slope<-windowData$rise / windowData$run

maxSlope<-aggregate(slope ~ chromID ,windowData, max)
colnames(maxSlope)<-c("chromID", "maxslope")
windowData<-merge(windowData, maxSlope, all = T)
windowData$normSlope<-windowData$slope / windowData$maxslope


maxcumulative<-aggregate(cumulative ~ chromID ,windowData, max)
colnames(maxcumulative)<-c("chromID", "maxcumsum")
windowData<-merge(windowData, maxcumulative, all = T)
windowData$normcumsum<-windowData$cumulative / windowData$maxcumsum


ggplot(windowData)+
  geom_rect(aes(xmin = Start, xmax =End, ymin = 0, ymax = normcumsum, fill = normSlope ))+
  # geom_line(aes(x = center, y = normcumsum, color = normSlope))+
  scale_fill_viridis()+
  facet_wrap(chromID ~ ., scales = "free")

ggplot(windowData)+geom_point(aes(x = normSlope, y = normcumsum))


#LOCAL MINIMA AND MAXIMA from https://gist.github.com/jamiefolson/5831746
which.peaks <- function(x,partial=TRUE,decreasing=FALSE){
  if (decreasing){
    if (partial){
      which(diff(c(TRUE,diff(x)<=0,FALSE))>0)
    }else {
      which(diff(diff(x)<=0)>0)
    }    
  }else {
    if (partial){
      which(diff(c(TRUE,diff(x)>=0,FALSE))<0)
    }else {
      which(diff(diff(x)>=0)<0)
    }
    
  }
}
x<- test
test<-c(1,2,3,45,5,59,1,2,2,2,23,2,1,1,1,12,1)
which.peaks(test, decreasing = TRUE)


minima<-data.frame()
for(c in unique(windowData$chromID)) {
  t <- windowData[windowData$chromID == c,]
  minima<- rbind(minima, t[which.peaks(t$SpenceRegsWeightMean, decreasing = FALSE ),c("center","chromID")]) #find valleys
}
minima$center_pos<-sqrt(minima$center ^ 2)

ggplot(windowData)+
  geom_rect(aes(xmin = Start, xmax =End, ymin = 0, ymax = SpenceRegsWeightMean, fill = normSlope ))+
  # geom_vline(data = minima, aes(xintercept = center_pos))+
  scale_fill_viridis()+
  facet_wrap(chromID ~ ., scales = "free")


# LOCAL MINIMA AND MAXIMA FROM AVERY
find_peaks <- function (x, m = 3){
  # from stack OVerflow! https://stats.stackexchange.com/questions/22974/how-to-find-local-peaks-valleys-in-a-series-of-data
  # available on github:https://github.com/stas-g/findPeaks
  # smaller m = more peaks; larger m = fewer peaks
  shape <- diff(sign(diff(x, na.pad = FALSE)))
  pks <- sapply(which(shape < 0), FUN = function(i){
    z <- i - m + 1
    z <- ifelse(z > 0, z, 1)
    w <- i + m + 1
    w <- ifelse(w < length(x), w, length(x))
    if(all(x[c(z : i, (i + 2) : w)] <= x[i + 1])) return(i + 1) else return(numeric(0))
  })
  pks <- unlist(pks)
  pks
}

idplotzones<-function(sampinfo,exampco,mychr,cenlens,posopt){
  # Finds chromosome "zones" (between local minima of CO density) for one chromosome (mychr); plots where these are. By combining all samples in sampinfo's COs
  # sampinfo, data.table containing the names of samples to read in (by substituting into exampco). Column names infname, usename
  # exampco, Example filestem for crossover files (suffixed with _<1-22, one per chromosome>.txt) with crossovers including location information (must include columns chr, pos.L, and pos.R). Where the sample name goes, should have _sampname_ exactly, including underscores.
  # mychr, chromosome to process. Number only!
  # cenlens, locations of centromeres and length of chromosomes. Columns: chr (format chr#), lcen, rcen, len
  # posopt, 'med', 'left', or 'right' -  COs are processed as having one location (for density), should midpoint or left/right boundary be used?
  
  # get in data
  allcos<-do.call(rbind,lapply(sampinfo$infname,function(x) fread.all(gsub("_sampname_",x,exampco),mychr,gsub("_sampname_",x,exampbcs),posopt)))
  d.allcos<-density(t)
  t<-c(1,1,1,2,3,4,6,5,7,8,9,1,2,4,6,6,3,1,1,1,5,5,6,7,4,3,4,3,2,1,9,9,9,9,9,9,9,9,9)
  # identify and format zones
  minima<-find_peaks(-d.allcos$y)
  myzones<-data.table(name=paste("chr",1,".",1:(length(minima)+1),sep=""),chr=paste("chr",1,sep=""),start=c(1,d.allcos$x[minima]+1),
                      end=c(d.allcos$x[minima],16))
  
  # plot (density with CEN with zones overlaid)
  plot(d.allcos,xlab=paste("Position on chromosome",1),main=paste("Chromosome",1,"zones"),ylim=c(0,max(d.allcos$y)))
  # for(i in 1:nrow(myzones)){ # sometimes a forloop is just easier
  #   if(i%%2>0){ # plot odds one color
  #     rect(myzones[i,start],0,myzones[i,end],max(d.allcos$y),col=rgb(255,0,0,100,max=255),border=NA)
  #   }else{ # plot evens another color
  #     rect(myzones[i,start],0,myzones[i,end],max(d.allcos$y),col=rgb(0,0,255,100,max=255),border=NA) #density=5,angle=45
  #   }
  # }
  # rect(cenlens[chr==paste("chr",mychr,sep=""),lcen],0,cenlens[chr==paste("chr",mychr,sep=""),rcen],max(d.allcos$y),col=rgb(190,190,190,150,max=255),border=NA)
  # rect(cenlens[chr==paste("chr",mychr,sep=""),lcen],0,cenlens[chr==paste("chr",mychr,sep=""),rcen],max(d.allcos$y),density=8,angle=45)
  # lines(d.allcos,lwd=2)
  # plotplain<-recordPlot() # Record to return
  # invisible(dev.off())
  # 
  ggplot()+geom_line(aes(x = d.allcos$x, y = d.allcos$y))+geom_rect(data=myzones,aes(xmin = start, xmax = end, ymin = 0, ymax = max(d.allcos$y), alpha = 0.1 ))

```

## Exploration of variables
## Multiple linear regression
## Location along chromosome arm


# Fine-scale patterns

## Description of variables

```{r finescaleSetup}

# Setup
today <- 20220622
gatfiledir<-paste0("analysis/",today,"_LocationPatterns/GATfiles/" )

# Frequency threshold
freq <- 0.2

# Directory
dir.create(gatfiledir, recursive = T)
```

### Segments

```{r makeSegments}

# Add libraries
pacman::p_load(vcfR, ggplot2)

# Setup 
dir.create(paste0(gatfiledir,"/Segments"))

# Read  VCF info
vcf<-read.vcfR("data/InversionGenotypes_processed/allgenotypes.vcf")

vcfInfo<-vcfR2tidy(vcf, info_only = TRUE)$fix

# Recalculate frquency for del vars
delinvs<-sub("Del", "",grep("Del", vcfInfo$ID, value =  T) )
delinvs<-delinvs[delinvs != "HsInv0030"]

for (var in delinvs) {
  
vartable<-vcfInfo[vcfInfo$ID == var, c("AC_EUR","AN_EUR")]
deltable<-vcfInfo[vcfInfo$ID == paste0(var, "Del"), c("AC_EUR","AN_EUR")]

AN_total<-(as.numeric(vartable$AN_EUR) - as.numeric(vartable$AC_EUR)) + as.numeric(vartable$AC_EUR) + as.numeric(deltable$AC_EUR)

vcfInfo[vcfInfo$ID == var, "AF_EUR"]<-as.character(as.numeric(vartable$AC_EUR)/AN_total)
}

vcfInfo$AF_EUR<-as.numeric(vcfInfo$AF_EUR)

# SELECT A FREQUENCY IN THE SETUP
# Comparison of selectefd frequency with distribution

ggplot(vcfInfo, aes(x = AF_EUR))+geom_histogram()+
  geom_vline(xintercept = summary(vcfInfo$AF_EUR)[3], color = "green")+
  geom_vline(xintercept=summary(vcfInfo$AF_EUR)[4], color = "blue")+
  geom_vline(xintercept = freq)

# Chromosome format
vcfInfo$CHROM<-paste0("chr",vcfInfo$CHROM)

# Classification by high-low frequency
lowfreq<-vcfInfo[vcfInfo$AF_EUR < freq,c("CHROM", "POS", "END", "ID")]
hifreq<-vcfInfo[vcfInfo$AF_EUR >= freq,c("CHROM", "POS", "END", "ID")]

write.table(lowfreq,paste0(gatfiledir,"/Segments/LFinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(hifreq, paste0(gatfiledir,"/Segments/HFinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(hifreq[hifreq$ID != "HsInv0501",], paste0(gatfiledir,"/Segments/HFno501invs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Classification by type
NH<-vcfInfo[ vcfInfo$TYPE == "NH",c("CHROM", "POS", "END", "ID")]
NAHR<-vcfInfo[ vcfInfo$TYPE == "NAHR",c("CHROM", "POS", "END", "ID")]

write.table(NH,paste0(gatfiledir,"/Segments/NHinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR, paste0(gatfiledir,"/Segments/NAHRinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR[NAHR$ID != "HsInv0501" ,], paste0(gatfiledir,"/Segments/NAHRno501invs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Classification by freq AND type
LFNH<-vcfInfo[vcfInfo$AF_EUR < freq & vcfInfo$TYPE == "NH",c("CHROM", "POS", "END", "ID")]
LFNAHR<-vcfInfo[vcfInfo$AF_EUR < freq & vcfInfo$TYPE == "NAHR",c("CHROM", "POS", "END", "ID")]
HFNH<-vcfInfo[vcfInfo$AF_EUR >= freq & vcfInfo$TYPE == "NH",c("CHROM", "POS", "END", "ID")]
HFNAHR<-vcfInfo[vcfInfo$AF_EUR >= freq & vcfInfo$TYPE == "NAHR",c("CHROM", "POS", "END", "ID")]
HFNAHRno<-vcfInfo[vcfInfo$AF_EUR >= freq & vcfInfo$TYPE == "NAHR" & vcfInfo$ID != "HsInv0501",c("CHROM", "POS", "END", "ID")]

write.table(LFNH, paste0(gatfiledir,"/Segments/LFNHinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(LFNAHR, paste0(gatfiledir,"/Segments/LFNAHRinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(HFNH, paste0(gatfiledir,"/Segments/HFNHinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(HFNAHR, paste0(gatfiledir,"/Segments/HFNAHRinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(HFNAHRno, paste0(gatfiledir,"/Segments/HFNAHRno501invs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# All invs
ALL<-vcfInfo[,c("CHROM", "POS", "END", "ID")]
ALLno<-vcfInfo[vcfInfo$ID != "HsInv0501",c("CHROM", "POS", "END", "ID")]

write.table(ALL, paste0(gatfiledir,"/Segments/invs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(ALLno, paste0(gatfiledir,"/Segments/no501invs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
```


### Annotations

```{bash makeAnnotarions}
ENVPY="/home/rgomez/anaconda3/bin/python"
# ANALYSIS WIDOWS: 5 fixed windows, and crossover zones with 0 persistence and 40k simulations (sample size similar to Bell et al.2020)

# Spence CEU map
$ENVPY code/python/recMapAnnotation.py -l data/SpenceSong_hg19_recMaps_processed/CEU_recombination_map_hg19_allChr.bed -p 0.001,0.01,0.1,0.999,0.99,0.9 -o analysis/$(date +'%Y%m%d')_LocationPatterns/ -d "CEUSpence"

# Bherer average
$ENVPY code/python/recMapAnnotation.py -l data/Bherer_Refined_genetic_map_b37_processed/sexavg_allChr.bed  -p 0.001,0.01,0.1,0.999,0.99,0.9 -o analysis/$(date +'%Y%m%d')_LocationPatterns/ -d "avgBherer"

# Bherer female
$ENVPY code/python/recMapAnnotation.py -l data/Bherer_Refined_genetic_map_b37_processed/female_allChr.bed -p 0.001,0.01,0.1,0.999,0.99,0.9 -o analysis/$(date +'%Y%m%d')_LocationPatterns/ -d "femBherer"


```

### Isochores

```{bash makeIsochores}
ENVPY="/home/rgomez/anaconda3/bin/python"

$ENVPY code/python/makeIsochores.py -f data/genomicSuperDups_coords.bed -t 0.09  -o "analysis/$(date +'%Y%m%d')_LocationPatterns" -d genomicSuperDups

```

### Workspace

```{bash makeWorkspace}
TODAY=$(date +'%Y%m%d')
USE=20220513

mkdir -p "analysis/$(date +'%Y%m%d')_LocationPatterns/GATfiles/Workspace/"

grep -v "cen" "analysis/${USE}_LocationPatterns/divideChromosomes/CEUSpence_COzones_0.1_800000/workspace.txt" | awk -v OFS="\t" 'NR>1{print $4,$1,$2}' >  "analysis/$(date +'%Y%m%d')_LocationPatterns/GATfiles/Workspace/workspace_CEUSpence.bed"

grep -v "cen" "analysis/${USE}_LocationPatterns/divideChromosomes/avgBherer_COzones_0.1_800000/workspace.txt" | awk -v OFS="\t" 'NR>1{print $4,$1,$2}' >  "analysis/$(date +'%Y%m%d')_LocationPatterns/GATfiles/Workspace/workspace_avgBherer.bed"

grep -v "cen" "analysis/${USE}_LocationPatterns/divideChromosomes/femBherer_COzones_0.1_800000/workspace.txt" | awk -v OFS="\t" 'NR>1{print $4,$1,$2}' >  "analysis/$(date +'%Y%m%d')_LocationPatterns/GATfiles/Workspace/workspace_femBherer.bed"


```


## GAT run

For generation methods

```{bash}

DESC="DefResults"
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
CONDITIONS="avgBherer repeats"
SEGMENTS="NHinvs NAHRnoOutinvs HFNHinvs LFNHinvs LFNAHRnoOutinvs HFNAHRnoOutinvs"
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv

for SEGMENT in $SEGMENTS; do
  #SEGNAME=$(echo $SEGMENT | rev | cut -d'/' -f 1 | rev | sed "s/\.bed//g")
  SEGNAME=$SEGMENT
  for NAME in $CONDITIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/Isochores_25Mwin_4cat.bed \
     --annotations=$FILESDIR/Annotations/annotations_${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_avgBherer.bed \
     --num-samples=5000 --log=$OUTDIR/log/${SEGNAME}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGNAME}_${NAME}.tsv

  done
done


for NAME in $CONDITIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done
```

I also want to test for repeat enrichment to see if isochores are working

```{bash}

DESC="Isochores_calculated"
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
CONDITIONS="avgBherer"
SEGMENTS="NHinvs NAHRnoOutinvs"
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv

for SEGMENT in $SEGMENTS; do
  SEGNAME=$SEGMENT
  for NAME in $CONDITIONS; do
 
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/isochores_calculated.bed \
     --annotations=$FILESDIR/Annotations/genomicSuperDups_coords.bed\
     --workspace=$FILESDIR/Workspace/workspace_${NAME}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGNAME}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGNAME}_${NAME}.tsv

  done
done
# <!--     --isochores=$FILESDIR/Isochores/isochores_calculated.bed \  -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $CONDITIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```

## GAT results - repeats


```{bash}
DESC="WorkspaceHigh0.1_noX"
TODAY=$(date +'%Y%m%d')
USE=20220620
CONDITIONS="CEUSpence avgBherer femBherer"
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $CONDITIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
pacman::p_load(reshape2, ggplot2, gridExtra)
desc<-"Isochores_calculated"
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("avgBherer")

figscale <- 3

for (condition in conditions){
  gatResultMerged <- read.table(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".tsv"), sep = "\t", header = T)
  
  # To add stars
  gatResultMerged$pstars<-ifelse(gatResultMerged$pvalue <= 0.05, ifelse(gatResultMerged$pvalue <= 0.01, ifelse(gatResultMerged$pvalue<= 0.001, "***", "**"), "*" ),"")
  gatResultMerged$qstars<-ifelse(gatResultMerged$qvalue <= 0.05, ifelse(gatResultMerged$qvalue <= 0.01, ifelse(gatResultMerged$qvalue<= 0.001, "***", "**"), "*" ),"")
  
  gatResultMerged$starpos<-ifelse(gatResultMerged$observed< gatResultMerged$CI95high,gatResultMerged$CI95high,gatResultMerged$observed )
  
  gatResultMerged<-merge(gatResultMerged,aggregate(starpos~ annotation, gatResultMerged, "max")  , by = "annotation", all.x = TRUE, suffixes = c(".simple", ".aggregated"))
  
  plot1<-ggplot(gatResultMerged, aes(x = track, y = l2fold))+ 
          geom_bar( stat="identity")+
          geom_text(aes(x = track, y = max(l2fold), label = pstars), color ="blue")+
          geom_text(aes(x = track, y = max(l2fold)+1, label = qstars), color = "red")+
          facet_wrap(annotation  ~ . )+
          scale_colour_manual(name="Significance",values=c("pvalue", "qvalue"), palette = c("blue", "red"))+
          ylab("log2( Fold change )")+
          xlab("Segment tracks")+
       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))
  
  plot2<- ggplot(gatResultMerged)+
    geom_pointrange(aes(x = track, y = expected, ymin=CI95low, ymax = CI95high))+
    geom_point(aes(x = track, y = observed) , shape = 4, size = 3)+
            geom_text(aes(x = track, y = starpos.aggregated*1.1, label = pstars), color ="blue")+
          geom_text(aes(x = track, y =  starpos.aggregated*1.2, label = qstars), color = "red")+
    facet_wrap(annotation  ~ ., scales = "free")+
     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ylab("Annotations overlappifng with segments\n(Nucleotide counts)")+
          xlab("Segment tracks")+
          ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))

      ggsave(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".png"),  grid.arrange(plot1, plot2), width=2480, height=3508, units = "px" , scale = 1.5)
}
```


## GAT results


```{bash}

TODAY=$(date +'%Y%m%d')
USE=$TODAY
CONDITIONS="CEUSpence avgBherer femBherer"
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults

for NAME in $CONDITIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
pacman::p_load(reshape2, ggplot2, gridExtra)

date<-"20220517"
conditions <-c("avgBherer", "femBherer", "CEUSpence")
figscale <- 3

for (condition in conditions){
  gatResultMerged <- read.table(paste0("analysis/",date,"_LocationPatterns/GATresults/",condition,".tsv"), sep = "\t", header = T)
  
  # To add stars
  gatResultMerged$pstars<-ifelse(gatResultMerged$pvalue <= 0.05, ifelse(gatResultMerged$pvalue <= 0.01, ifelse(gatResultMerged$pvalue<= 0.001, "***", "**"), "*" ),"")
  gatResultMerged$qstars<-ifelse(gatResultMerged$qvalue <= 0.05, ifelse(gatResultMerged$qvalue <= 0.01, ifelse(gatResultMerged$qvalue<= 0.001, "***", "**"), "*" ),"")
  
  gatResultMerged$starpos<-ifelse(gatResultMerged$observed< gatResultMerged$CI95high,gatResultMerged$CI95high,gatResultMerged$observed )
  
  gatResultMerged<-merge(gatResultMerged,aggregate(starpos~ annotation, gatResultMerged, "max")  , by = "annotation", all.x = TRUE, suffixes = c(".simple", ".aggregated"))
  
  plot1<-ggplot(gatResultMerged, aes(x = track, y = l2fold))+ 
          geom_bar( stat="identity")+
          geom_text(aes(x = track, y = max(l2fold), label = pstars), color ="blue")+
          geom_text(aes(x = track, y = max(l2fold)+1, label = qstars), color = "red")+
          facet_wrap(annotation  ~ . )+
          scale_colour_manual(name="Significance",values=c("pvalue", "qvalue"), palette = c("blue", "red"))+
          ylab("log2( Fold change )")+
          xlab("Segment tracks")+
       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))
  
  plot2<- ggplot(gatResultMerged)+
    geom_pointrange(aes(x = track, y = expected, ymin=CI95low, ymax = CI95high))+
    geom_point(aes(x = track, y = observed) , shape = 4, size = 3)+
            geom_text(aes(x = track, y = starpos.aggregated*1.1, label = pstars), color ="blue")+
          geom_text(aes(x = track, y =  starpos.aggregated*1.2, label = qstars), color = "red")+
    facet_wrap(annotation  ~ ., scales = "free")+
     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ylab("Annotations overlappifng with segments\n(Nucleotide counts)")+
          xlab("Segment tracks")+
          ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))

      ggsave(paste0("analysis/",date,"_LocationPatterns/GATresults/",condition,".png"),  grid.arrange(plot1, plot2), width=2480, height=3508, units = "px" , scale = 1.5)
}
```

# Analysis: are invesions in regions with higher repeats?

```{r}
library(ggplot2)
rep<-read.table("data/genomicSuperDups_coords.bed", header = T)
inv<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NAHRinvs.bed", header = F)
invNH<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NHinvs.bed", header = F)
invNAHR<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NAHRinvs.bed", header = F)
win<-read.table("analysis/20220622_LocationPatterns/GATfiles/Isochores/genomicSuperDups_COzones_0.09/windows.txt")
colnames(win)<-c("chrom", "start", "end", "type")

colnames(inv)<-colnames(invNH)<-colnames(invNAHR)<-colnames(rep)

rep$center<-(rep$chromStart+rep$chromEnd)/2

chromnames<- paste0("chr", c(0:22))
ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  geom_rect(data =win, aes(xmin = start, xmax = end, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = invNH, aes(xintercept = (chromStart + chromEnd)/2 ), color = "blue", alpha = 0.5)+
  geom_vline(data = invNAHR, aes(xintercept = (chromStart + chromEnd)/2 ), color = "red", alpha = 0.5)




inv$density<-
  apply(inv, 1, function(x){
       invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
  d<-density(rep[as.character(rep$chrom) == x["chrom"],"center"])
  val<-mean(approx(d$x, d$y, xout = c(invStart:invEnd))$y)
  val
} )

 
 inv$nucOverlap<-
   apply(inv, 1, function(x){
     invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })
 
 invNAHR<-inv
# ---
 rep<-read.table("data/genomicSuperDups_coords.bed", header = T)
inv<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NHinvs.bed", header = F)

colnames(inv)<-colnames(rep)

rep$center<-(rep$chromStart+rep$chromEnd)/2

chromnames<- paste0("chr", c(0:22))
ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 ), color = "red", alpha = 0.5)


density(rep[rep$chrom %in% chromnames,"center"])

 inv$density<-
  
  apply(inv, 1, function(x){
       invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
  d<-density(rep[as.character(rep$chrom) == x["chrom"],"center"])
  val<-mean(approx(d$x, d$y, xout = c(invStart:invEnd))$y)
  val
} )

  inv$nucOverlap<-
   apply(inv, 1, function(x){
     invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })
 
invNH<-inv

# --

invNAHR$Type<-"NAHR"
invNH$Type<-"NH"
inv<-rbind(invNAHR, invNH)
# inv<-inv[!(inv[,4] %in% c("HsInv0501", "HsInv382", "HsInv786", "HsInv1110", "HsInv0573")),]
ggplot(inv)+geom_boxplot(aes(x = Type, y = density))

ggplot(inv)+geom_boxplot(aes(x = Type, y = log10(nucOverlap)))

ggplot(inv)+geom_boxplot(aes(x = Type, y = (nucOverlap/(chromEnd - chromStart)  )))

pacman::p_load("Hmisc", "reshape2")

summary((nucOverlap ) ~ Type, data = inv, fun=mean)

hist(log10(inv$nucOverlap))
```


```{r}
#OK let's try to find out regions with similar repeat density

workspace<-win<-read.table("analysis/20220622_LocationPatterns/GATfiles/Workspace/workspace_avgBherer.bed")
colnames(workspace)<-c("chrom", "start", "end")

# Which is the ideal win size? - 15k
# prob<-invNAHR[invNAHR$nucOverlap > 50000, ]
# min(prob$chromEnd - prob$chromStart)

# I divide the genome in 15k windows
size <-  2.5 * 1000000
genoWindows<-do.call("rbind", apply(workspace, 1, function(x){
  v<-seq(x["start"], as.numeric(x["end"])+size, size)
  starts<-v[1:length(v)-1]
  ends<-v[2:length(v)]
  data.frame(x["chrom"], starts, ends)
}))

colnames(genoWindows)<-colnames(workspace)

# Now I calculate the repeat nucleotide overlap for each of them
genoWindows$nucOverlap<-   apply(genoWindows, 1, function(x){
     invStart <- as.numeric(x["start"])
     invEnd<- as.numeric(x["end"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })

#See
   
ggplot()+
  # geom_point(data = genoWindows[genoWindows$nucOverlap > 2272,], aes(x = (end+start)/2, y = nucOverlap) , size = 1, alpha = 0.2)+
  geom_rect(data = genoWindows, aes(xmin = start, xmax = end, ymin = 0 , ymax = nucOverlap) ,  alpha = 0.2)+
  geom_hline(yintercept = 2e+06)+
  # geom_rect(data =isochores_newcomplete, aes(xmin = regstart, xmax = regend, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_wrap(chrom ~ ., scales = "free_x")+
   geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 , color = problematic, linetype = Type))+
   geom_rect(data = inv, aes(xmin = chromStart, xmax = chromEnd, ymin = 0 , ymax = Inf, fill = problematic), alpha = 0.2)+
  ggtitle("test")
```


```{r}
# I put a threshold and calculate high repeat regions
winSize<-2.5*1000000
genoWindows$nucDensity <- genoWindows$nucOverlap / winSize



threshold1<-0.05 # test 06
# threshold<-7e+06
# inv$problematic<-ifelse(inv$nucOverlap >30000& inv$nucOverlap<150000, TRUE, FALSE )

genoWindows$isochore<-ifelse(genoWindows$nucDensity >= threshold1, "High", "Low")
genoWindows[genoWindows$nucDensity >= 0.05, "isochore"]<- "High005"
# genoWindows[genoWindows$nucDensity >= 0.25, "isochore"]<- "High025"
genoWindows[genoWindows$nucDensity >= 0.8, "isochore"]<- "High08"
genoWindows[genoWindows$nucDensity >= 2.8, "isochore"]<- "High28"
  
totest<-inv[inv$Type=="NAHR" &( inv$nucOverlap <30000), ]


ggplot()+
  # geom_point(data = genoWindows[genoWindows$nucOverlap > 2272,], aes(x = (end+start)/2, y = nucOverlap) , size = 1, alpha = 0.2)+
  geom_rect(data = genoWindows, aes(xmin = start, xmax = end, ymin = 0 , ymax = nucOverlap) ,  alpha = 0.5)+
  geom_rect(data = genoWindows, aes(xmin = start, xmax = end, ymin = 0 , ymax = Inf, fill = isochore) ,  alpha = 0.2)+
  geom_hline(yintercept = threshold1, alpha= 0.5)+
  # geom_rect(data =isochores_newcomplete, aes(xmin = regstart, xmax = regend, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_grid(chrom ~ ., scales = "free_x")+
   geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 , color = problematic, linetype = Type))+
  # geom_vline(data = totest, aes(xintercept = (chromStart + chromEnd)/2 , linetype = Type))+
   # geom_rect(data = inv, aes(xmin = chromStart, xmax = chromEnd, ymin = 0 , ymax = Inf, fill = problematic), alpha = 0.2)+
  ggtitle("test")


write.table(genoWindows[,c("chrom", "start", "end", "isochore")], paste0(gatfiledir,"/Isochores/isochores_calculated.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(totest[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/NAHRnoOutinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

hflist<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/HFNAHRinvs.bed")
write.table(totest[totest[,4] %in% hflist$V4 ,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HFNAHRnoOutinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(totest[!(totest[,4] %in% hflist$V4),c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LFNAHRnoOutinvs.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

```


```{r}
# What is hapending in the problemanics?
theps<-inv[ inv$chrom != "chrX" & inv$chrom != "chrY",]

theps<-cbind(theps, do.call("rbind",apply(theps, 1, function(x){
  point<-(as.numeric(x["chromStart"])+as.numeric(x["chromEnd"]))/2
  df<-genoWindows[as.character(genoWindows$chrom) == x["chrom"] &  genoWindows$start <= point & genoWindows$end >= point ,c("nucDensity", "isochore" )]
  # ifelse(nrow(df)>0, df, data.frame(nucOverlap = NA, isochore= NA))
})))


b<-ggplot(theps,aes(x = repdensity, y  =nucDensity))+geom_point(aes( color = isochore, shape = problematic))+geom_smooth(method = lm)
a
b
```


```{r newIsochores}
inv$repdensity<-inv$nucOverlap / (inv$chromEnd - inv$chromStart)

invNAHR$repdensity<-invNAHR$nucOverlap / (invNAHR$chromEnd - invNAHR$chromStart)
invNH$repdensity<-invNH$nucOverlap / (invNH$chromEnd - invNH$chromStart)

# highreg<-inv[inv$repdensity > mean(invNH$repdensity),] # test 1
# highreg<-invNAHR #test2
# highreg<-invNAHR[invNAHR[,4] != "HsInv0501",] #test3
# highreg<-invNAHR[!(invNAHR[,4] %in% c("HsInv0501", "HsInv382", "HsInv786", "HsInv1110", "HsInv0573")) & invNAHR$repdensity > mean(invNH$repdensity),] #test4
highreg<-invNAHR[(invNAHR[,4] %in% c("HsInv0486", "HsInv0374", "HsInv1117")),] #test5
# highreg<-inv[inv$nucOverlap > 50000,]

highreg$regstart<-highreg$chromStart - 25000
highreg$regend<-highreg$chromStart + 25000

highreg[highreg$regstart <0, "regstart"] <- 0

isochore_new<-unique(highreg[order(highreg$regstart),c("chrom", "regstart", "regend")])
isochore_new$type<-"High"

isochores_newcomplete<-data.frame()
for(c in unique(win$chrom)){
  cstart<-min(win[win$chrom == c, "start"])
  cend<-max(win[win$chrom == c, "end"])

  subs<-isochore_new[isochore_new$chrom == c,]
  
  sts<-subs$regstart
  eds<-subs$regend
  if (nrow(subs)>1){
    
    for( i in c(2:length(sts))){
      if(sts[i]<eds[i-1]){
        sts[i]<-NA
        eds[i-1]<-NA
        } 
    }
  }
  
    pos<-unique(c(cstart,cend,sts, eds))
    pos<-pos[!is.na(pos) & pos <= cend]
    pos<-pos[order(pos)]
    subs<-data.frame(chrom=c, regstart=pos[1:length(pos)-1], regend=pos[2:length(pos)], type = "Low", stringsAsFactors = F)
    subs[subs$regstart %in% sts, "type"]<-"High" 
    
    isochores_newcomplete<-rbind(isochores_newcomplete, subs)

  }
  

ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  geom_rect(data =isochores_newcomplete, aes(xmin = regstart, xmax = regend, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = invNH, aes(xintercept = (chromStart + chromEnd)/2 ), color = "blue", alpha = 0.5)+
  geom_vline(data = invNAHR, aes(xintercept = (chromStart + chromEnd)/2 ), color = "red", alpha = 0.5)


write.table(isochores_newcomplete, paste0(gatfiledir,"/Isochores/isochores_calculated.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

```


```{r finescaleSetup}

# Setup
today <- 20220621
gatfiledir<-paste0("analysis/",today,"_LocationPatterns/GATfiles/" )

# Frequency threshold
freq <- 0.2

# Directory
dir.create(gatfiledir, recursive = T)
```

```{r}

workspace<-data.frame(chrom = "chrA", start = 0, end = 1000)
isochores<-data.frame(chrom = "chrA", start = c(0, 500), end = c(500, 1000), type = c("High", "Low"))
annotations<-data.frame(chrom = "chrA", start  = c(seq(5, 500, 25), seq(505, 1000, 50)) )
annotations$end<-annotations$start + 5
segments<-data.frame(chrom = "chrA", start = seq(1,1000, 100))
segments$end<-segments$start + 50
segments$type<-ifelse(segments$start<500, "NAHR", "NH")


write.table(workspace, paste0(gatfiledir,"/Workspace/workspace_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(isochores, paste0(gatfiledir,"/Isochores/isochores_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(annotations, paste0(gatfiledir,"/Annotations/annotations_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(segments[,c("chrom", "start", "end")], paste0(gatfiledir,"/Segments/segments_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(segments[segments$type == "NAHR",], paste0(gatfiledir,"/Segments/segmentsNAHR_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(segments[segments$type == "NH",], paste0(gatfiledir,"/Segments/segmentsNH_sim.bed"), quote=F, col.names = F, sep = "\t", row.names = F)



ggplot()+geom_rect(data = annotations,aes(xmin = start, xmax = end, ymin = -1, ymax =1 ))+
  geom_rect(data = segments, aes(xmin = start, xmax = end, ymin = -1, ymax =1 , fill = type), alpha = 0.5)+
  geom_rect(data = isochores, aes(xmin = start, xmax = end, ymin = -1, ymax = 1, color  = type), alpha = 0.1)



```


```{bash}

DESC="Simulation"
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
CONDITIONS="sim"

mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv

for SEGMENT in $(ls $FILESDIR/Segments/*_sim.bed); do
  SEGNAME=$(echo $SEGMENT | rev | cut -d'/' -f 1 | rev | sed "s/\.bed//g")
  for NAME in $CONDITIONS; do
    
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$SEGMENT --isochores=$FILESDIR/Isochores/isochores_${NAME}.bed\
     --annotations=$FILESDIR/Annotations/genomicSuperDups_coords.bed \
     --workspace=$FILESDIR/Workspace/workspace_${NAME}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGNAME}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGNAME}_${NAME}.tsv

  done
done
# <!-- --     --isochores=$FILESDIR/Isochores/isochores_${NAME}.bed \-->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $CONDITIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```

```{r}
pacman::p_load(reshape2, ggplot2, gridExtra)
desc<-"Simulation"
date<-"20220622"
conditions <-c("sim")
figscale <- 3

for (condition in conditions){
  gatResultMerged <- read.table(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".tsv"), sep = "\t", header = T)
  
  # To add stars
  gatResultMerged$pstars<-ifelse(gatResultMerged$pvalue <= 0.05, ifelse(gatResultMerged$pvalue <= 0.01, ifelse(gatResultMerged$pvalue<= 0.001, "***", "**"), "*" ),"")
  gatResultMerged$qstars<-ifelse(gatResultMerged$qvalue <= 0.05, ifelse(gatResultMerged$qvalue <= 0.01, ifelse(gatResultMerged$qvalue<= 0.001, "***", "**"), "*" ),"")
  
  gatResultMerged$starpos<-ifelse(gatResultMerged$observed< gatResultMerged$CI95high,gatResultMerged$CI95high,gatResultMerged$observed )
  
  gatResultMerged<-merge(gatResultMerged,aggregate(starpos~ annotation, gatResultMerged, "max")  , by = "annotation", all.x = TRUE, suffixes = c(".simple", ".aggregated"))
  
  plot1<-ggplot(gatResultMerged, aes(x = track, y = l2fold))+ 
          geom_bar( stat="identity")+
          geom_text(aes(x = track, y = max(l2fold), label = pstars), color ="blue")+
          geom_text(aes(x = track, y = max(l2fold)+1, label = qstars), color = "red")+
          facet_wrap(annotation  ~ . )+
          scale_colour_manual(name="Significance",values=c("pvalue", "qvalue"), palette = c("blue", "red"))+
          ylab("log2( Fold change )")+
          xlab("Segment tracks")+
       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))
  
  plot2<- ggplot(gatResultMerged)+
    geom_pointrange(aes(x = track, y = expected, ymin=CI95low, ymax = CI95high))+
    geom_point(aes(x = track, y = observed) , shape = 4, size = 3)+
            geom_text(aes(x = track, y = starpos.aggregated*1.1, label = pstars), color ="blue")+
          geom_text(aes(x = track, y =  starpos.aggregated*1.2, label = qstars), color = "red")+
    facet_wrap(annotation  ~ ., scales = "free")+
     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ylab("Annotations overlappifng with segments\n(Nucleotide counts)")+
          xlab("Segment tracks")+
          ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))

      ggsave(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".png"),  grid.arrange(plot1, plot2), width=2480, height=3508, units = "px" , scale = 1.5)
}
```


```{r}
inv$size <- inv$chromEnd - inv$chromStart

inv$problematic<-ifelse(inv$nucOverlap > 50000, TRUE, FALSE)


ggplot(inv)+geom_point(aes(x = log10(size), y =log10(nucOverlap), color = problematic))


ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  geom_rect(data =isochores_newcomplete, aes(xmin = regstart, xmax = regend, fill = type, ymin = 0, ymax = Inf) , alpha = 0.3)+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 , color = problematic))


```
# A LIMPIO 

```{r finescaleSetup}

# Setup
today <- 20220622
gatfiledir<-paste0("analysis/",today,"_LocationPatterns/GATfiles/" )

# Frequency threshold
freq <- 0.2

# Directory
dir.create(gatfiledir, recursive = T)
```

Useful function for later: (GAT plots)
```{r}
pacman::p_load(reshape2, ggplot2, gridExtra)

GATplots<-function(desc, date, conditions){

figscale <- 3

for (condition in conditions){
  gatResultMerged <- read.table(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".tsv"), sep = "\t", header = T)
  
  # To add stars
  gatResultMerged$pstars<-ifelse(gatResultMerged$pvalue <= 0.05, ifelse(gatResultMerged$pvalue <= 0.01, ifelse(gatResultMerged$pvalue<= 0.001, "***", "**"), "*" ),"")
  gatResultMerged$qstars<-ifelse(gatResultMerged$qvalue <= 0.05, ifelse(gatResultMerged$qvalue <= 0.01, ifelse(gatResultMerged$qvalue<= 0.001, "***", "**"), "*" ),"")
  
  gatResultMerged$starpos<-ifelse(gatResultMerged$observed< gatResultMerged$CI95high,gatResultMerged$CI95high,gatResultMerged$observed )
  
  gatResultMerged<-merge(gatResultMerged,aggregate(starpos~ annotation, gatResultMerged, "max")  , by = "annotation", all.x = TRUE, suffixes = c(".simple", ".aggregated"))
  
  plot1<-ggplot(gatResultMerged, aes(x = track, y = l2fold))+ 
          geom_bar( stat="identity")+
          geom_text(aes(x = track, y = max(l2fold), label = pstars), color ="blue")+
          geom_text(aes(x = track, y = max(l2fold)+1, label = qstars), color = "red")+
          facet_wrap(annotation  ~ . )+
          scale_colour_manual(name="Significance",values=c("pvalue", "qvalue"), palette = c("blue", "red"))+
          ylab("log2( Fold change )")+
          xlab("Segment tracks")+
       theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))
  
  plot2<- ggplot(gatResultMerged)+
    geom_pointrange(aes(x = track, y = expected, ymin=CI95low, ymax = CI95high))+
    geom_point(aes(x = track, y = observed) , shape = 4, size = 3)+
            geom_text(aes(x = track, y = starpos.aggregated*1.1, label = pstars), color ="blue")+
          geom_text(aes(x = track, y =  starpos.aggregated*1.2, label = qstars), color = "red")+
     geom_text(aes(x = track, y =  starpos.aggregated*1.3, label = paste0("N=",track_nsegments )), color = "black")+
    facet_wrap(annotation  ~ ., scales = "free")+
     theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+
            ylab("Annotations overlappifng with segments\n(Nucleotide counts)")+
          xlab("Segment tracks")+
          ggtitle(paste0("Summarized GAT result: Inversions vs. ERRs in ",condition))

      ggsave(paste0("analysis/",date,"_LocationPatterns/GATresults/",desc,"/",condition,".png"),  grid.arrange(plot1, plot2), width=2480, height=3508, units = "px" , scale = 1.5)
}
}
```


### The invs info con los repeats

```{r}
library(ggplot2)
rep<-read.table("data/genomicSuperDups_coords.bed", header = T)
inv<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NAHRinvs.bed", header = F)
invNH<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NHinvs.bed", header = F)
invNAHR<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NAHRinvs.bed", header = F)
win<-read.table("analysis/20220622_LocationPatterns/GATfiles/Isochores/genomicSuperDups_COzones_0.09/windows.txt")
colnames(win)<-c("chrom", "start", "end", "type")

colnames(inv)<-colnames(invNH)<-colnames(invNAHR)<-colnames(rep)

rep$center<-(rep$chromStart+rep$chromEnd)/2

chromnames<- paste0("chr", c(0:22))
ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  geom_rect(data =win, aes(xmin = start, xmax = end, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = invNH, aes(xintercept = (chromStart + chromEnd)/2 ), color = "blue", alpha = 0.5)+
  geom_vline(data = invNAHR, aes(xintercept = (chromStart + chromEnd)/2 ), color = "red", alpha = 0.5)




inv$density<-
  apply(inv, 1, function(x){
       invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
  d<-density(rep[as.character(rep$chrom) == x["chrom"],"center"])
  val<-mean(approx(d$x, d$y, xout = c(invStart:invEnd))$y)
  val
} )

 
 inv$nucOverlap<-
   apply(inv, 1, function(x){
     invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })
 
 invNAHR<-inv
# ---
 rep<-read.table("data/genomicSuperDups_coords.bed", header = T)
inv<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/NHinvs.bed", header = F)

colnames(inv)<-colnames(rep)

rep$center<-(rep$chromStart+rep$chromEnd)/2

chromnames<- paste0("chr", c(0:22))
ggplot()+geom_density(data = rep[rep$chrom %in% chromnames,], aes(center))+
  facet_wrap(chrom ~ ., scales = "free_x")+
  geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 ), color = "red", alpha = 0.5)


density(rep[rep$chrom %in% chromnames,"center"])

 inv$density<-
  
  apply(inv, 1, function(x){
       invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
  d<-density(rep[as.character(rep$chrom) == x["chrom"],"center"])
  val<-mean(approx(d$x, d$y, xout = c(invStart:invEnd))$y)
  val
} )

  inv$nucOverlap<-
   apply(inv, 1, function(x){
     invStart <- as.numeric(x["chromStart"])
     invEnd<- as.numeric(x["chromEnd"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })
 
invNH<-inv

# --

invNAHR$Type<-"NAHR"
invNH$Type<-"NH"
inv<-rbind(invNAHR, invNH)
# inv<-inv[!(inv[,4] %in% c("HsInv0501", "HsInv382", "HsInv786", "HsInv1110", "HsInv0573")),]
ggplot(inv)+geom_boxplot(aes(x = Type, y = density))

ggplot(inv)+geom_boxplot(aes(x = Type, y = log10(nucOverlap)))

ggplot(inv)+geom_boxplot(aes(x = Type, y = (nucOverlap/(chromEnd - chromStart)  )))

pacman::p_load("Hmisc", "reshape2")

summary((nucOverlap ) ~ Type, data = inv, fun=mean)

hist(log10(inv$nucOverlap))
```


### NH invs analysis por un lado , estas no tienen problema en principio #WORK

```{bash}
DESC="NHinvs_comparable_i"
SEGMENTS="NHinvs HFNHinvs LFNHinvs"
ISOCHORES="isochores_NAHR_All_comparable"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY="20220622"
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
#  ### ----#


for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \
     --output-counts-pattern=$OUTDIR/counts/${SEGMENT}_${NAME}%s.tsv.gz\
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
desc<-"NHinvs_comparable_i"
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```


### Un poco de análisis de las NAHR invs 

No hay diferencias entre HF y LF para las isocoras

```{r}
colnames(invNAHR)[4]<-"ID"
invNAHR$nucDensity<-with(invNAHR,  nucOverlap / ( chromEnd- chromStart ) )

hflist<-read.table("analysis/20220622_LocationPatterns/GATfiles/Segments/HFinvs.bed")

invNAHR$freq<-ifelse(invNAHR$ID %in% hflist[,4], "HF", "LF")

summary(nucDensity ~ freq, invNAHR)

ggplot(invNAHR)+geom_boxplot(aes(x = freq, y = nucDensity))

```

Relacion significativa  logtamaño-densidad de repeats

```{r}
ggplot(invNAHR, aes(x = log10(chromEnd- chromStart), y = nucDensity, color= freq))+geom_point()+geom_smooth(method = "lm")
invNAHR$size<-with(invNAHR, chromEnd- chromStart)
invNAHR$logsize<-log10(invNAHR$size)
GGally::ggpairs(  data = invNAHR, columns = c(8, 11),  ggplot2::aes(colour=freq))

m<-lm(nucDensity ~ logsize , invNAHR)

summary(m)

library("ggpubr")
ggscatter(invNAHR, x = "logsize", y = "nucDensity", 
          add = "reg.line", conf.int = TRUE, 
          cor.coef = TRUE, cor.method = "pearson")
```



### NAHR that have the same mean nucDensity as NH #WOKR, little sample

```{r}
colnames(inv)[4]<-"ID"
inv$nucDensity<-with(inv,  nucOverlap / ( chromEnd- chromStart ) )

# ggplot(inv)+geom_boxplot(aes(x = Type, y = log10(nucOverlap)))

NAHR_NHlike<-invNAHR[invNAHR$nucOverlap <= mean((invNH$nucOverlap)),]


write.table(NAHR_NHlike[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/NAHR_NHlike.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_NHlike[NAHR_NHlike$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HFNAHR_NHlike.bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_NHlike[NAHR_NHlike$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LFNAHR_NHlike.bed"), quote=F, col.names = F, sep = "\t", row.names = F)

```


```{bash}

DESC="NAHR_NHlike"
SEGMENTS="NAHR_NHlike HFNAHR_NHlike LFNAHR_NHlike"
#ISOCHORES="isochores_calculated"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
DESC="NAHR_NHlike"
desc<-"NAHR_NHlike"
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```


### NAHR <=0.47 rep Density #WORK little sample

```{r}
# colnames(inv)[4]<-"ID"
# inv$nucDensity<-with(inv,  nucOverlap / ( chromEnd- chromStart ) )

# ggplot(inv)+geom_boxplot(aes(x = Type, y = log10(nucOverlap)))

NAHR_05<-invNAHR[invNAHR$nucDensity <=0.47,]

name<-"NAHR_047"
write.table(NAHR_05[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)

```


```{bash}

DESC="NAHR_047"
SEGMENTS="${DESC} HF${DESC} LF${DESC}"
#ISOCHORES="isochores_calculated"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}

desc<-name
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```



### Make repeat density for isochores

I make 2.5Mb windows
```{r}
#OK let's try to find out regions with similar repeat density

workspace<-win<-read.table("analysis/20220622_LocationPatterns/GATfiles/Workspace/workspace_avgBherer.bed")
colnames(workspace)<-c("chrom", "start", "end")

# Which is the ideal win size? - 15k
# prob<-invNAHR[invNAHR$nucOverlap > 50000, ]
# min(prob$chromEnd - prob$chromStart)

# I divide the genome in 15k windows
size <-  2.5 * 1000000
genoWindows<-do.call("rbind", apply(workspace, 1, function(x){
  v<-seq(x["start"], as.numeric(x["end"])+size, size)
  starts<-v[1:length(v)-1]
  ends<-v[2:length(v)]
  data.frame(x["chrom"], starts, ends)
}))

colnames(genoWindows)<-colnames(workspace)

# Now I calculate the repeat nucleotide overlap for each of them
genoWindows$nucOverlap<-   apply(genoWindows, 1, function(x){
     invStart <- as.numeric(x["start"])
     invEnd<- as.numeric(x["end"])
     repC<-rep[as.character(rep$chrom) == x["chrom"],]
     repO<-unique(repC[( (  repC$chromEnd %in% c(invStart:invEnd)  )  | 
             (  repC$chromStart %in% c(invStart:invEnd) ) | 
             (  repC$chromStart < invStart&  repC$chromEnd > invEnd )
            ), ])
     repO[repO$chromStart < invStart , "chromStart"]<-invStart
     repO[repO$chromEnd > invEnd , "chromEnd"]<-invEnd
     
     repO$size<-repO$chromEnd - repO$chromStart
     sum(repO$size)
   })

#See
   
ggplot()+
  # geom_point(data = genoWindows[genoWindows$nucOverlap > 2272,], aes(x = (end+start)/2, y = nucOverlap) , size = 1, alpha = 0.2)+
  geom_rect(data = genoWindows, aes(xmin = start, xmax = end, ymin = 0 , ymax = nucOverlap) ,  alpha = 0.2)+
  geom_hline(yintercept = 2e+06)+
  # geom_rect(data =isochores_newcomplete, aes(xmin = regstart, xmax = regend, fill = type, ymin = 0, ymax = Inf) , alpha = 0.5)+
  facet_wrap(chrom ~ ., scales = "free_x")+
   geom_vline(data = inv, aes(xintercept = (chromStart + chromEnd)/2 , color = problematic, linetype = Type))+
   geom_rect(data = inv, aes(xmin = chromStart, xmax = chromEnd, ymin = 0 , ymax = Inf, fill = problematic), alpha = 0.2)+
  ggtitle("test")
```

Function to make isochores
```{r}
# I put a threshold and calculate high repeat regions
isoCreator<-function(densList, name){

winSize<-2.5*1000000
genoWindows$nucDensity <- genoWindows$nucOverlap / winSize

genoWindows$isochore<-"Low"
for(i in densList){
  genoWindows[genoWindows$nucDensity >= i, "isochore"]<- paste0("High",sub("\\.", "", i))
}
  
write.table(genoWindows[,c("chrom", "start", "end", "isochore")], paste0(gatfiledir,"/Isochores/isochores_",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
}

isoCreatorOve<-function(densList, name){

winSize<-2.5*1000000
genoWindows$nucDensity <- genoWindows$nucOverlap / winSize

genoWindows$isochore<-"Low"
for(i in densList){
  genoWindows[genoWindows$nucOverlap >= i, "isochore"]<- paste0("High",sub("\\.", "", i))
}
  
write.table(genoWindows[,c("chrom", "start", "end", "isochore")], paste0(gatfiledir,"/Isochores/isochores_",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
}
```


### Test isochores

#### BigInvs
```{r}
#Set inversion to test
NAHR_05<-invNAHR[invNAHR$nucDensity >=3.9,]

name<-"NAHR_Big"
write.table(NAHR_05[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Set isochores to test
isoCreator(c(2.8,0.8), name)

# Plot old available


```


```{bash}

DESC="NAHR_Big"
SEGMENTS="${DESC} HF${DESC} LF${DESC}"
ISOCHORES="isochores_${DESC}"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
desc<-name
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```


#### MediumInvs
```{r}
#Set inversion to test
NAHR_05<-invNAHR[invNAHR$nucDensity <= 1.2 | invNAHR$nucDensity >= 2.5  ,]
# NAHR_05<-invNAHR[invNAHR$nucDensity >= 2.5 ,]

name<-"NAHR_Med"
write.table(NAHR_05[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Set isochores to test
isoCreator(c( 0.05, 0.8,1.8, 2.8), name)

# Plot old available


```


```{bash}

DESC="NAHR_Med"
SEGMENTS="${DESC} HF${DESC} LF${DESC}"
ISOCHORES="isochores_${DESC}"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY=$(date +'%Y%m%d')
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
desc<-name
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```


#### AllInvs
```{r}
#Set inversion to test
NAHR_05<-invNAHR[invNAHR$nucDensity <= 1.6 | invNAHR$nucDensity >= 1.7  ,]
# NAHR_05<-invNAHR[invNAHR$nucDensity >= 2.5 ,]

name<-"NAHR_All"
write.table(NAHR_05[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Set isochores to test
isoCreator(c( 0.05, 0.8,1.6, 2.8), name)

# Plot old available


```


```{bash}

DESC="NAHR_All"
SEGMENTS="${DESC} HF${DESC} LF${DESC}"
ISOCHORES="isochores_${DESC}"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY="20220622"
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
desc<-name
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```



#### AllInvs_toCompare

```{r}
#Set inversion to test
NAHR_05<-invNAHR[invNAHR$nucDensity <= 1.6 | invNAHR$nucDensity >= 1.7  ,]
# NAHR_05<-invNAHR[invNAHR$nucDensity >= 2.5 ,]

name<-"NAHR_All_comparable"
write.table(NAHR_05[,c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "HF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/HF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)
write.table(NAHR_05[NAHR_05$freq == "LF",c("chrom", "chromStart", "chromEnd")], paste0(gatfiledir,"/Segments/LF",name,".bed"), quote=F, col.names = F, sep = "\t", row.names = F)

# Set isochores to test
isoCreator(c( 0.05, 0.8,1.6, 2.8), name)

# Plot old available


```


```{bash}

DESC="NAHR_All_comparable"
SEGMENTS="${DESC} HF${DESC} LF${DESC}"
ISOCHORES="isochores_${DESC}"
WORKSPACE="avgBherer"
ANNOTATIONS="annotations_repeats annotations_avgBherer"

# NO TOCAR ### ----#
TODAY="20220622"
USE=$TODAY
FILESDIR=analysis/${USE}_LocationPatterns/GATfiles
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}
mkdir -p  $OUTDIR/log $OUTDIR/counts $OUTDIR/tsv
# -------###--------#

for SEGMENT in $SEGMENTS; do
  for NAME in $ANNOTATIONS; do
     /home/rgomez/anaconda3/bin/gat-run.py  \
     --segments=$FILESDIR/Segments/${SEGMENT}.bed \
     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \
     --annotations=$FILESDIR/Annotations/${NAME}.bed\
     --output-counts-pattern=$OUTDIR/counts/${SEGMENT}_${NAME}%s.tsv.gz\
     --workspace=$FILESDIR/Workspace/workspace_${WORKSPACE}.bed \
     --num-samples=1000 --log=$OUTDIR/log/${SEGMENT}_${NAME}.log  --stdout=$OUTDIR/tsv/${SEGMENT}_${NAME}.tsv

  done
done
# <!--     #     --isochores=$FILESDIR/Isochores/${ISOCHORES}.bed \ -->
OUTDIR=analysis/${TODAY}_LocationPatterns/GATresults/${DESC}

for NAME in $ANNOTATIONS; do
  HEADERFILE=$(ls $OUTDIR/tsv/* |grep $NAME |head -n1)
  head -n1 $HEADERFILE > $OUTDIR/$NAME.tsv
  for FILE in $(ls $OUTDIR/tsv/* |grep $NAME); do
    TRACK=$(echo $FILE | rev |cut -d"/" -f1 |rev |cut -d "_" -f1)
    tail -n+2 $FILE | sed "s/merged/${TRACK}/g" >> $OUTDIR/$NAME.tsv
  done
done

```


```{r}
desc<-name
date<-"20220622"
# conditions <-c("avgBherer", "repeats")
conditions <-c("annotations_repeats" ,"annotations_avgBherer")

GATplots(desc, date, conditions)
```

# General plot ONE FOR ALL

```{r}

names<-c("NAHR_All", "NHinvs")
feature<-"annotations_avgBherer.tsv"

tab<-data.frame()

for (n in names){
  tab<-rbind(tab,read.table(paste0("analysis/20220622_LocationPatterns/GATresults/",n, "/", feature), header = T))
}

tab$freq<-substr(tab$track, 0, 2)
tab$type<-gsub("HF|LF|invs", "", tab$track)
  
tab$pstars<-ifelse(tab$pvalue <= 0.05, ifelse(tab$pvalue <= 0.01, ifelse(tab$pvalue<= 0.001, "***", "**"), "*" ),"")  
tab$qstars<-ifelse(tab$qvalue <= 0.05, ifelse(tab$qvalue <= 0.01, ifelse(tab$qvalue<= 0.001, "***", "**"), "*" ),"")
# levels(tab$type)<-c("NAHR", "NH")
tab$type<-ordered(tab$type, levels = c("NH" ,"NAHR"))
tab$alphaval<-as.numeric(as.character(factor(tab$pstars, levels =c("", "*", "**", "***"), labels =c( 0.2, 1, 1, 1 ))))
tab$Frequency<-ifelse(tab$freq == "HF", "High", "Low")
c<-ggplot(tab[tab$freq %in% c("HF", "LF"),])+geom_bar(aes(y= l2fold , x  = annotation , fill = Frequency, alpha = alphaval), stat = "identity", position = "dodge")+
  geom_text(aes(x = annotation, y = l2fold, label = pstars, group = freq, color = Frequency), position = position_dodge(width = .9), angle = 90)+
  scale_x_discrete(  label = c("<0.1%", "<1%", "<10%", ">90%", ">99%", ">99.9%") )+
  xlab("")+ylab("Fold change")+
  facet_grid(.~type)+
   scale_alpha_identity()
  

ggsave(paste0("hotspots.png"),  c, width=2480, height=3508/3, units = "px" , scale = 1)


```


